src/dask_awkward/lib/optimize.py

```diff
@@ -97,166 +60,96 @@ def _projectable_io_layer_names(dsk: HighLevelGraph) -> list[str]:
-    # final necessary key is the 0th partition of the last layer in
-    # the graph (hence the toposort to find last layer).
-    final_key = (dsk._toposort_layers()[-1], 0)
-    new_hlg = HighLevelGraph(layers, deps).cull([final_key])
-    get_sync(new_hlg, list(new_hlg.keys()))
-
-
-def _necessary_columns_brute(dsk: HighLevelGraph) -> dict:
-    # staring fields should be those belonging to the AwkwardIOLayer's
-    # metadata (typetracer) array.
-    out_meta = list(dsk.layers.values())[-1]._meta  # type: ignore
-    keep = out_meta.layout.form.columns()
-    columns: list[str] = []
-
-    pio_layer_names = _projectable_io_layer_names(dsk)
-    if len(pio_layer_names) > 1:
-        raise RuntimeError(
-            "'brute' method of optimization currently only graphs with a single IO layer."
-        )
-    pio_layer = pio_layer_names[0]
-
-    columns = dsk.layers[pio_layer]._meta.layout.form.columns()  # type: ignore
-
-    # can only select output columns that exist in the input
-    # (other names may have come from aliases)
-    keep = [c for c in out_meta.layout.form.columns() if c in columns]
-
-    for c in columns:
-        if c in keep:
-            continue
-        holdout = c
-        allcolumns = set(columns)
-        remaining = list(allcolumns - {holdout})
-        try:
-            _attempt_compute_with_columns_brute(dsk, columns=remaining)
-        except IndexError:
-            keep.append(holdout)
-    if keep == columns:
-        keep = None
-    return {pio_layer: keep}
-
-
-def _has_projectable_awkward_io_layer(dsk: HighLevelGraph) -> bool:
-    for _, v in dsk.layers.items():
-        if isinstance(v, AwkwardIOLayer) and hasattr(v.io_func, "project_columns"):
-            return True
-    return False
-
-
-def optimize_iolayer_columns_brute(dsk: HighLevelGraph) -> HighLevelGraph:
-    # if the task graph doesn't contain a column-projectable
-    # AwkwardIOLayer then bail on this optimization (just return the
-    # existing task graph).
-    if not _has_projectable_awkward_io_layer(dsk):
-        return dsk
-    # determine the necessary columns to complete the executation of
-    # the metadata (typetracer) based task graph.
-    necessary_cols = _necessary_columns_brute(dsk)
-    layer_name, necessary_cols = list(_necessary_columns_brute(dsk).items())[0]
+def _output_io_layer_names(dsk: HighLevelGraph) -> list[str]:
+    return [
+        n
+        for n, v in dsk.layers.items()
+        if (v.annotations or {}).get("ak_output")
+    ]
 
-    if necessary_cols is None:
-        return dsk  # type: ignore
+def _mock_io_func(*args, **kwargs):
+    import awkward as ak
+    for arg in args + tuple(kwargs.values()):
+        if isinstance(arg, ak.Array):
+            arg.layout._touch_data(recursive=True)
+
+
+def _mock_output(layer: Layer):
+    assert len(layer.dsk) == 1
+    layer2 = copy.deepcopy(layer)
+    mp = layer2.mapping.copy()  # why is this always a MetrializedLayer?
+    key = iter(mp).__next__()
```

Comment:
I've been playing around with this today. When an output layer is present if we don't modify each partition's write step the optimization will fail because `dask_awkward.lib.io.parquet._write_partition` is still getting called on a typetracer.

For example, if we have 3 partitions we'll end up with a graph that looks like

```python
{
    (write-parquet-abc123, 0): (_mock_io_func, ..., 0),
    (write-parquet-abc123, 1): (_ToParquetFn, ..., 1),  # <-- raises b/c callable is getting a typetracer
    (write-parquet-abc123, 2): (_ToParquetFn, ..., 2),
}
```

```suggestion
    for k in iter(mp.keys()):
        mp[k] = (_mock_output_func,) + mp[k][1:]
```

Reply:
OK, right - I think this was meant to go in conjunction with the usual route of selecting just one output key, but of course you can't have both those code paths at once.